server/src/main/java/com/linecorp/line/auth/fido/fido2/server/attestation/tpm/TpmParser.java

```diff
@@ -119,50 +143,77 @@ public static PubArea parsePubArea(byte[] pubArea) throws IOException {
+        byte[] keyBitsBytes = new byte[2];
+        byte[] exponentBytes = new byte[4];
+
+        inputStream.read(symmetricBytes);
+        inputStream.read(schemeBytes);
+        inputStream.read(keyBitsBytes);
+        inputStream.read(exponentBytes);
 
-        // read unique (key value)
+        Parameters parameters = new RsaParameters();
+        parameters.setSymmetric(symmetricBytes);
+        parameters.setScheme(TpmSignatureAlgorithm.fromValue(UnsignedUtil.readUINT16BE(schemeBytes)));
+        ((RsaParameters) parameters).setKeyBits(keyBitsBytes);
+        ((RsaParameters) parameters).setExponent(exponentBytes);
+        return parameters;
+    }
+
+    private static byte[] extractUniqueBytesForRSA(ByteArrayInputStream inputStream) throws IOException {
+        byte[] uniqueLengthBytes = new byte[2];
         inputStream.read(uniqueLengthBytes);
         int uniqueLength = UnsignedUtil.readUINT16BE(uniqueLengthBytes);
-        uniqueBytes = new byte[uniqueLength];
+        byte[] uniqueBytes = new byte[uniqueLength];
         inputStream.read(uniqueBytes);
+        return uniqueBytes;
+    }
 
-        return PubArea
-                .builder()
-                .type(TpmKeyAlgorithm.fromValue(type))
-                .nameAlg(nameAlg)
-                .objectAttributes(objectAttributes)
-                .parameters(parameters)
-                .unique(uniqueBytes)
-                .build();
+    private static Parameters extractParametersForECC(ByteArrayInputStream inputStream) throws IOException {
+        byte[] symmetricBytes = new byte[2];
+        byte[] schemeBytes = new byte[2];
+        byte[] curveIdBytes = new byte[2];
+        byte[] kdfBytes = new byte[2];
+
+        inputStream.read(symmetricBytes);
+        inputStream.read(schemeBytes);
+        inputStream.read(curveIdBytes);
+        inputStream.read(kdfBytes);
+
+        Parameters parameters = new EccParameters();
+        parameters.setSymmetric(symmetricBytes);
+        parameters.setScheme(TpmSignatureAlgorithm.fromValue(UnsignedUtil.readUINT16BE(schemeBytes)));
+        ((EccParameters) parameters).setCurveId(TpmEccCurve.fromValue(UnsignedUtil.readUINT16BE(curveIdBytes)));
+        ((EccParameters) parameters).setKdf(kdfBytes);
+        return parameters;
+    }
+
+    private static byte[] extractUniqueBytesForECC(ByteArrayInputStream inputStream) throws IOException {
+        byte[] xLengthBytes = new byte[2];
+        inputStream.read(xLengthBytes);
+        int xLength = UnsignedUtil.readUINT16BE(xLengthBytes);
+        byte[] xBytes = new byte[xLength];
+        inputStream.read(xBytes);
+
+        byte[] yLengthBytes = new byte[2];
+        inputStream.read(yLengthBytes);
+        int yLength = UnsignedUtil.readUINT16BE(yLengthBytes);
+        byte[] yBytes = new byte[yLength];
+        inputStream.read(yBytes);
+
+        return ByteBuffer.allocate(xLength + yLength)
+                .put(xBytes)
+                .put(yBytes)
+                .array();
```

Comment:
This might be invalid in some cases: if the coordinate most significant byte is zero, some implementation might not encode it.

So you can end up in a cases where the X coordinate is encoded on 31 bytes and the Y coordinate on 32 bytes.

I'm not sure if this has an impact on the attestation verification however.

Reply:
I am _very_ familiar with the ASN.1 integer encoding issue @haxelion describes, but I don't believe it applies here.  If the encoded the byte array properly to spec, this code should extract it properly.

There is code in this [library here](https://github.com/line/line-fido2-server/blob/78f78c6cc50586560d35b85777a8fef71446d321/server/src/main/java/com/linecorp/line/auth/fido/fido2/server/util/SignatureUtil.java#L170) that exhibits this problem, however the referenced method appears to be unused and should probably be removed from the project to avoid confusion or anyone else accidentally resuing it.  There is an `convertASN1toXMLDSIG()` method in Java that implements this conversion correctly, I believe that is what this library is eventually and correctly using via one of the `java.security.Signature.verifySHA###withECDSA()` methods.

Reply:
@aseigler I only took a quick look through the TPM spec but I couldn't find any detail on how it handles coordinate encoding. Part 2, 11.2.5.2 TPMS_ECC_POINT does not give any details and I couldn't find more information in Part 1 annex C. That's why I'm unsure how implementation deal with a leading 00 byte. If you have any reference on those details I'm interested.

Reply:
It's out of scope, that's purely an ASN.1 encoding issue. TCG spec only goes as far as how a compliant TPM should encode the already ASN.1 encoded bytes into TCG structures.

Reply:
I see, so we can assume every leading zero bytes are preserved and all those buffers have the same size.